state/factory/workingset.go

```diff
@@ -591,22 +599,41 @@ func (ws *workingSet) pickAndRunActions(
 				actionIterator.PopAccount()
 				continue
 			}
-			// TODO: error may cause the tx to be stuck in the action pool until expired
 			receipt, err := ws.runAction(actionCtx, nextAction)
-			switch errors.Cause(err) {
-			case nil:
-				// do nothing
-			case action.ErrChainID:
-				continue
-			case action.ErrGasLimit, errUnfoldTxContainer:
-				actionIterator.PopAccount()
-				continue
-			default:
-				nextActionHash, hashErr := nextAction.Hash()
-				if hashErr != nil {
-					return nil, errors.Wrapf(hashErr, "Failed to get hash for %x", nextActionHash)
+			if fCtx.PurgeActpoolOnRunActionError {
+				switch errors.Cause(err) {
+				case nil:
+					// do nothing
+				case action.ErrChainID, action.ErrGasLimit, errUnfoldTxContainer, errDeployerNotWhitelisted:
```

Comment:
should we also call `ap.DeleteAction()` here?

Reply:
should remove from actpool

Reply:
but the ErrGasLimit action may run successfully in next block?

Reply:
not for `action.ErrGasLimit`